File: sim/antchain_sim/miner.py
"""Miner agent model: heterogeneous hashrate, ACO skill, and strategy."""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Literal

import numpy as np

from .tsp import AntColonyOptimizer, TSPInstance

Strategy = Literal["honest", "pheromone_hoarder", "withholder"]


@dataclass
class Miner:
    miner_id: int
    hashrate_share: float          # relative hashrate, sums to 1 across population
    aco_skill: float                # multiplier on effective ACO iterations per tick
    strategy: Strategy = "honest"
    is_strategic: bool = False      # convenience flag: strategy != "honest"

    # per-block mutable state
    optimizer: AntColonyOptimizer | None = field(default=None, repr=False)
    best_length: float = np.inf
    best_tour: np.ndarray | None = field(default=None, repr=False)
    _search_budget_accum: float = 0.0  # carries fractional ACO-iteration budget across ticks

    def start_block(self, instance: TSPInstance, rng: np.random.Generator) -> None:
        """Called at the start of every block round."""
        self._search_budget_accum = 0.0
        if self.optimizer is None or self.strategy != "pheromone_hoarder":
            self.optimizer = AntColonyOptimizer(instance, rng=rng)
        else:
            # pheromone_hoarder: keep tau from the previous block instead of
            # resetting it, carrying a (protocol-illegitimate) memory advantage
            # forward onto the new instance. Distance/heuristic table (eta)
            # must still be refreshed for the new instance.
            self.optimizer.instance = instance
            with np.errstate(divide="ignore"):
                self.optimizer.eta = np.where(instance.dist > 0, 1.0 / instance.dist, 0.0)
            self.optimizer.best_tour = None
            self.optimizer.best_length = np.inf
        self.best_length = np.inf
        self.best_tour = None

    def search_step(self, search_iterations: float) -> float:
        """Run ACO for a (possibly fractional) number of iterations this tick,
        scaled by aco_skill. Fractional budget below one full iteration is
        carried forward (not discarded) so low-hashrate-share miners still
        make search progress over many ticks instead of being rounded to
        zero every tick."""
        self._search_budget_accum += search_iterations * self.aco_skill
        n_iter = int(self._search_budget_accum)
        self._search_budget_accum -= n_iter
        if n_iter > 0 and self.optimizer is not None:
            best_tour, best_length = self.optimizer.step(n_iter)
            if best_tour is not None:
                self.best_tour, self.best_length = best_tour, best_length
        return self.best_length


def make_population(
    n_miners: int,
    strategic_fraction: float = 0.15,
    strategic_strategies: tuple[Strategy, ...] = ("pheromone_hoarder", "withholder"),
    hashrate_alpha: float = 1.5,
    skill_low: float = 0.6,
    skill_high: float = 1.8,
    seed: int = 0,
) -> list[Miner]:
    """Build a heterogeneous miner population.

    Hashrate shares are drawn from a Pareto-like heavy tail (mimicking real
    mining-pool concentration) and normalized to sum to 1. A configurable
    fraction of miners are "strategic" (pheromone hoarders or withholders);
    the rest are honest.
    """
    rng = np.random.default_rng(seed)
    raw = rng.pareto(hashrate_alpha, size=n_miners) + 1.0
    shares = raw / raw.sum()
    skills = rng.uniform(skill_low, skill_high, size=n_miners)

    n_strategic = int(round(n_miners * strategic_fraction))
    strategic_ids = set(rng.choice(n_miners, size=n_strategic, replace=False).tolist())

    miners = []
    for i in range(n_miners):
        if i in strategic_ids:
            strat = strategic_strategies[i % len(strategic_strategies)]
        else:
            strat = "honest"
        miners.append(
            Miner(
                miner_id=i,
                hashrate_share=float(shares[i]),
                aco_skill=float(skills[i]),
                strategy=strat,
                is_strategic=strat != "honest",
            )
        )
    return miners

File: cpp/src/miner.cpp
#include "miner.hpp"

#include <algorithm>
#include <cmath>
#include <numeric>

namespace antchain {

void Miner::start_block(const TSPInstance& inst, uint64_t block_seed) {
    search_budget_accum = 0.0;
    if (!optimizer || strategy != Strategy::PheromoneHoarder) {
        opt_seed = mix64(block_seed ^ (static_cast<uint64_t>(miner_id) * 0x9E3779B97F4A7C15ULL));
        optimizer = std::make_unique<AntColonyOptimizer>(inst, opt_seed);
    } else {
        // pheromone-hoarder: keep tau_ from the previous block, only rebind
        // the instance (distances change, "memory" doesn't reset)
        optimizer->rebind_instance_keep_pheromones(inst);
    }
    best_length = 1e18;
    best_tour.clear();
}

double Miner::search_step(double search_iterations) {
    search_budget_accum += search_iterations * aco_skill;
    int n_iter = static_cast<int>(search_budget_accum);
    search_budget_accum -= n_iter;
    if (n_iter > 0 && optimizer) {
        auto [tour, length] = optimizer->step(n_iter);
        if (!tour.empty()) {
            best_tour = tour;
            best_length = length;
        }
    }
    return best_length;
}

std::vector<Miner> make_population(int n_miners, double strategic_fraction, uint64_t seed,
                                    double hashrate_alpha, double skill_low, double skill_high) {
    SplitMix64 rng(seed);

    // numpy Generator.pareto(a): Lomax draw via inverse-CDF, X = (1-U)^(-1/a) - 1
    std::vector<double> raw(n_miners);
    double raw_sum = 0.0;
    for (int i = 0; i < n_miners; ++i) {
        double u = rng.next_double();
        double x = std::pow(1.0 - u, -1.0 / hashrate_alpha) - 1.0;
        raw[i] = x + 1.0;
        raw_sum += raw[i];
    }

    std::vector<double> skills(n_miners);
    for (int i = 0; i < n_miners; ++i) {
        skills[i] = skill_low + rng.next_double() * (skill_high - skill_low);
    }

    int n_strategic = static_cast<int>(std::lround(n_miners * strategic_fraction));
    // reservoir-free sample without replacement: partial Fisher-Yates over indices
    std::vector<int> idx(n_miners);
    std::iota(idx.begin(), idx.end(), 0);
    for (int i = 0; i < n_strategic && i < n_miners; ++i) {
        int j = i + rng.next_int(n_miners - i);
        std::swap(idx[i], idx[j]);
    }
    std::vector<bool> is_strategic(n_miners, false);
    for (int i = 0; i < n_strategic; ++i) is_strategic[idx[i]] = true;

    static const Strategy kStrategicStrategies[2] = {Strategy::PheromoneHoarder, Strategy::Withholder};

    std::vector<Miner> miners(n_miners);
    for (int i = 0; i < n_miners; ++i) {
        Miner& m = miners[i];
        m.miner_id = i;
        m.hashrate_share = raw[i] / raw_sum;
        m.aco_skill = skills[i];
        m.is_strategic = is_strategic[i];
        m.strategy = is_strategic[i] ? kStrategicStrategies[i % 2] : Strategy::Honest;
    }
    return miners;
}

} // namespace antchain
